core/src/main/scala/org/apache/spark/ui/jobs/StageTable.scala

```diff
@@ -145,9 +145,20 @@ private[ui] class StageTableBase(
       case None => "Unknown"
     }
     val finishTime = s.completionTime.getOrElse(System.currentTimeMillis)
-    val duration = s.submissionTime.map { t =>
-      if (finishTime > t) finishTime - t else System.currentTimeMillis - t
-    }
+    val taskLaunchTimes = stageData.taskData.values
+        .filter(_.taskInfo.launchTime != 0)
+        .map(_.taskInfo.launchTime)
```

Comment:
That is much clearer, thanks. And I'll add a comment.

I was guarding against an empty list of launch times, which could throw an exception if `min` were called on it.  Do I need to worry about that here?